code/controllers/subsystem/tts.dm

```diff
@@ -0,0 +1,219 @@
+	for(var/client/player in unfiltered_players)
+		if(!player?.prefs.read_preference(/datum/preference/toggle/sound_tts))
+			listeners -= player
+			continue
+
+		if(has_html_audio && player?.prefs.read_preference(/datum/preference/toggle/sound_tts_use_html_audio))
+			html_audio_players += player
+			listeners -= player
+
+	if(has_html_audio)
+		var/datum/asset_cache_item/cached_mp3 = new(identifier, filepath)
+		cached_mp3.namespace = "text_to_speech"
+		SSassets.transport.register_asset(identifier, cached_mp3)
+		var/current_text_url = SSassets.transport.get_asset_url(null, cached_mp3)
+		SShtml_audio.play_audio(current_text_url, html_audio_players)
+	var/sound/sound_to_play = sound(file(filepath))
+	for(var/client/player in listeners)
+		SEND_SOUND(player, sound_to_play)
+
+/datum/controller/subsystem/tts/proc/handle_request(list/entry)
+	var/time_left = entry[TIMEOUT_INDEX]
+	var/estimated_processing_time = estimate_word_processing_length(entry[MESSAGE_INDEX])
+	if(time_left < world.time || (time_left < world.time + message_timeout_early_minimum && time_left - estimated_processing_time < world.time))
+		cached_voices -= entry[IDENTIFIER_INDEX]
+		return
+	var/datum/http_request/request = entry[REQUEST_INDEX]
+	request.begin_async()
+	in_process_tts_messages += list(entry)
+
+/datum/controller/subsystem/tts/fire(resumed)
+	if(!tts_enabled)
+		flags |= SS_NO_FIRE
+		return
+
+	if(!resumed)
+		while(length(in_process_tts_messages) < max_concurrent_requests && queued_tts_messages.len > 0)
+			var/list/entry = popleft(queued_tts_messages)
+			handle_request(entry)
+		current_processing_tts_messages = in_process_tts_messages.Copy()
+
+	// For speed
+	var/list/processing_messages = current_processing_tts_messages
+	while(processing_messages.len)
+		var/current_message = processing_messages[processing_messages.len]
+		processing_messages.len--
+		if(current_message[TIMEOUT_INDEX] < world.time)
+			in_process_tts_messages -= list(current_message)
+			cached_voices -= current_message[IDENTIFIER_INDEX]
+			continue
+
+		var/datum/http_request/request = current_message[REQUEST_INDEX]
+		if(!request.is_complete())
+			continue
+
+		var/datum/http_response/response = request.into_response()
+		in_process_tts_messages -= list(current_message)
+		if(response.errored)
+			cached_voices -= current_message[IDENTIFIER_INDEX]
+			continue
+		var/identifier = current_message[IDENTIFIER_INDEX]
+		var/list/tts_targets = current_message[TARGET_INDEX]
+		for(var/list/targets in current_message[EXTRA_TARGETS_INDEX])
+			tts_targets |= targets
+		play_tts(identifier, "tmp/tts/[identifier].mp3", tts_targets)
+		cached_voices -= identifier
+		var/time_taken = (message_timeout - (current_message[TIMEOUT_INDEX] - world.time)) / 10
+		rtf = time_taken / estimate_word_processing_length(current_message[MESSAGE_INDEX], 1)
+		if(MC_TICK_CHECK)
+			return
+
+/datum/controller/subsystem/tts/proc/queue_tts_message(message, speaker, list/targets, filter)
```

Comment:
You're running this inside preferences act, so it should not yield. Please add SHOULD_NOT_SLEEP(TRUE) and make it work.

Reply:
This needs to yield. I'll just invoke it asynchronously from within preferences act.